source/compiler-core/slang-json-value.cpp

```diff
@@ -808,4 +808,318 @@ bool JSONContainer::areEqual(const JSONValue& a, const JSONValue& b)
     return false;
 }
 
+void JSONContainer::traverseRecursively(const JSONValue& value, JSONListener* listener)
+{
+    typedef JSONValue::Type Type;
+
+    switch (value.type)
+    {    
+        case Type::True:            return listener->addBoolValue(true, value.loc);
+        case Type::False:           return listener->addBoolValue(false, value.loc);
+        case Type::Null:            return listener->addNullValue(value.loc);
+
+        case Type::StringLexeme:    return listener->addLexemeValue(JSONTokenType::StringLiteral, getLexeme(value), value.loc);
+        case Type::IntegerLexeme:   return listener->addLexemeValue(JSONTokenType::IntegerLiteral, getLexeme(value), value.loc);
+        case Type::FloatLexeme:     return listener->addLexemeValue(JSONTokenType::FloatLiteral, getLexeme(value), value.loc);
+
+        case Type::IntegerValue:    return listener->addIntegerValue(value.intValue, value.loc); 
+        case Type::FloatValue:      return listener->addFloatValue(value.floatValue, value.loc);
+        case Type::StringValue:
```

Comment:
It's a little unfortunate to need these distinct cases for what amounts to "processed" and "unprocessed" values of the various leaf types. What would be the down-side to having the `JSONValue` type only support the `IntegerValue`, etc. cases and not the `IntegerLexeme` etc. cases?